### PR TITLE
Add program cache for executor.py

### PR DESCRIPTION
**Pull request by jacquesqiao**
related issue: https://github.com/PaddlePaddle/Paddle/issues/8729

### CODE REVIEW

File: python/paddle/fluid/executor.py [hunk 1]
```diff
@@ -225,7 +226,19 @@ def run(self,
             feed_var_name='feed',
             fetch_var_name='fetch',
             scope=None,
-            return_numpy=True):
+            return_numpy=True,
+            use_program_cache=False):
+        """
+        :param program: the program that need to run
```
Comment (panyx0718) [inline python/paddle/fluid/executor.py]: 这个是核心API，comment能多写点会更好，比如program是从头运行到尾，还是根据fetch的依赖关系只运行一部分

Comment (panyx0718) [reply python/paddle/fluid/executor.py]: 甚至可以在里面写个简单的python example

File: python/paddle/fluid/executor.py [hunk 2]
```diff
@@ -225,7 +226,19 @@ def run(self,
...
+        :param feed: feed variable list
+        :param fetch_list: fetch variable list
+        :param feed_var_name: feed_var_name default to 'feed'
```
Comment (panyx0718) [inline python/paddle/fluid/executor.py]: Maybe a bit more comment here? I'm confused about the difference between "fetch_list" and "fetch_var_name". And the meaning of default 'feed'/'fetch'.

File: python/paddle/fluid/executor.py [hunk 3]
```diff
@@ -225,7 +226,19 @@ def run(self,
             feed_var_name='feed',
             fetch_var_name='fetch',
             scope=None,
-            return_numpy=True):
+            return_numpy=True,
+            use_program_cache=False):
```
Comment (panyx0718) [inline python/paddle/fluid/executor.py]: I would remove this 'option' and use 'True' by default. Then as a quick follow up, auto-detect the program change (preferred) and auto invalidate cache.

File: python/paddle/fluid/executor.py [hunk 4]
```diff
@@ -240,35 +253,64 @@ def run(self,
...
+                    assert isinstance(var, Variable) or isinstance(var, str), (
+                        "Wrong type for fetch_list[%s]: %s" % (i, type(var)))
+                    global_block.append_op(
+                        type='fetch',
+                        inputs={'X': [var]},
+                        outputs={'Out': [fetch_var]},
+                        attrs={'col': i})
+
+        # feed var to framework
+        for op in program_cache.global_block().ops:
```
Comment (panyx0718) [inline python/paddle/fluid/executor.py]: could program_cache be None if use_program_cache=False?

File: python/paddle/fluid/executor.py [hunk 5]
```diff
@@ -225,7 +226,19 @@ def run(self,
...
+        :param program: the program that need to run
+        :param feed: feed variable list
+        :param fetch_list: fetch variable list
+        :param feed_var_name: feed_var_name default to 'feed'
+        :param fetch_var_name: fetch_var_name default to 'fetch'
+        :param scope: the scope used to run this program, you can switch it to different scope.
+        :param return_numpy: convert the fetched tensor to numpy
+        :param use_program_cache: set use_program_cache to true if program not changed compare to the last step.
+        :return:
+        """
```
Comment (panyx0718) [inline python/paddle/fluid/executor.py]: more return comment?

File: python/paddle/fluid/executor.py [hunk 6]
```diff
@@ -240,35 +253,64 @@ def run(self,
...
-                out = global_block.var(name)
-                global_block.prepend_op(
-                    type='feed',
-                    inputs={'X': [feed_var]},
-                    outputs={'Out': [out]},
-                    attrs={'col': i})
-
-        for op in global_block.ops:
+            if use_program_cache:
+                self.program_caches[program_cache_key] = program_cache
```
Comment (panyx0718) [inline python/paddle/fluid/executor.py]: Maybe limit the number cached here? Save memory if the program change often.

**Comment by dzhwinter** [comment]
BTW, in fact, our `feed`, `fetch` operator does nothing, we use `fetch_feed_method` of pybind instead. Is there any profiling result about these two operators?
Maybe we should delete the `feed`, `fetch` ASAP.

File: python/paddle/fluid/executor.py (hunk 1 above)
Comment (jacquesqiao) [reply python/paddle/fluid/executor.py]: done

File: python/paddle/fluid/executor.py (hunk 2 above)
Comment (jacquesqiao) [reply python/paddle/fluid/executor.py]: done

File: python/paddle/fluid/executor.py (hunk 3 above)
Comment (jacquesqiao) [reply python/paddle/fluid/executor.py]: currently, the cache cannot check if the program has changed, so user need to set this flag if they are sure the program has not changed, so we have to set this to false by default.

File: python/paddle/fluid/executor.py [hunk 7]
```diff
@@ -177,6 +193,7 @@ def __init__(self, places):
         # TODO(dzhwinter) : only use the first place
         self.executor = core.Executor(act_places[0])
         self.places = places
+        self.program_caches = dict()
```
Comment (panyx0718) [inline python/paddle/fluid/executor.py]: self._program_caches?

File: python/paddle/fluid/executor.py [hunk 8]
```diff
@@ -240,35 +278,64 @@ def run(self,
...
+        if use_program_cache:
+            # find program cache by cache_key
+            program_cache = self.program_caches.get(program_cache_key, None)
+            # TODO(qiao): Should check program_cache and program are exactly the same.
         else:
-            feed_var = global_block.create_var(
-                name=feed_var_name,
-                type=core.VarDesc.VarType.FEED_MINIBATCH,
-                persistable=True)
+            self.program_caches.pop(program_cache_key, None)
```
Comment (panyx0718) [inline python/paddle/fluid/executor.py]: why pop here?

File: python/paddle/fluid/executor.py [hunk 9]
```diff
@@ -225,9 +242,30 @@ def run(self,
...
+        :param fetch_var_name: the name for the output variable of feed Operator.
+        :param scope: the scope used to run this program, you can switch it to different scope. default is global_scope
+        :param return_numpy: if convert the fetched tensor to numpy
+        :param use_program_cache: set use_program_cache to true if program not changed compare to the last step.
+        :return: result according to fetch_list.
+        """
         if feed is None:
             feed = {}
+        if not isinstance(feed, dict):
+            raise TypeError("feed should be a map")
```
Comment (panyx0718) [inline python/paddle/fluid/executor.py]: It seems feed can be a list before? Will this break existing user? But I guess it's OK if list doesn't make sense here.

File: python/paddle/fluid/executor.py [hunk 10]
```diff
@@ -240,35 +278,64 @@ def run(self,
...
-                    inputs={'X': [feed_var]},
-                    outputs={'Out': [out]},
-                    attrs={'col': i})
-
-        for op in global_block.ops:
+        if program_cache is None:
+            program_cache = program.clone()
+
+            if use_program_cache:
+                self.program_caches[program_cache_key] = program_cache
```
Comment (panyx0718) [inline python/paddle/fluid/executor.py]: Leave a TODO here to avoid caching too many programs?

**Pull request merged**